src/thunderbots/software/ai/navigator/path_planner/astar.h

```diff
@@ -0,0 +1,155 @@
+    // vertex type - a 2D point on a grid
+    typedef GridGraph2D::vertex_descriptor GridVertex;
+
+    // a hash function for a grid vertex, necessary for all the maps
+    // that boost::astar_search uses
+    struct grid_vertex_hash : std::unary_function<GridVertex, std::size_t>
+    {
+        std::size_t operator()(const GridVertex &gp) const;
+    };
+
+    class AStarGridGraph
+    {
+       public:
+        AStarGridGraph() = delete;
+
+        /**
+         * Constructs a grid graph from a Field object.
+         *
+         * @param field the field to create a graph for
+         * @param _grid_vertex_density how many vertices on the grid graph, per metre
+         */
+        explicit AStarGridGraph(const Field &field, const size_t _grid_vertex_density);
+
+        /**
+         * Converts a grid point to a Point on the field.
+         *
+         * @param grid_v the grid vertex to convert to a Point.
+         */
+        Point gridPointToPoint(const GridVertex &grid_v);
+
+        /**
+         * Finds the nearest vertex on the grid graph to the
+         * given point.
+         *
+         * @param point the point to find a grid vertex close to it
+         */
+        GridVertex nearestGridVertex(const Point &point);
+
+        /**
+         * Returns a const reference to the grid graph itself.
+         * @return the underlying grid graph.
+         */
+        const GridGraph2D &graph();
+
+        /**
+         * The distance between vertices on the grid graph.
+         * @return the distance between vertices on the grid graph.
+         */
+        constexpr double gridVertexDistance();
+
+       private:
+        std::unique_ptr<GridGraph2D> field_graph;
+
+        // this map is used to convert from the grid points that A* will
+        // return, into Points that can be used for navigation
+        std::vector<std::pair<Point, GridVertex>> grid_points_list;
+
+        // the minimum and maximum x- and y-values on the field
+        const double field_min_x, field_min_y;
+        const size_t grid_vertex_density;
+    };
+
+    class AStarHeuristic : public boost::astar_heuristic<GridGraph2D, edge_cost_t>
+    {
+       public:
+        AStarHeuristic() = delete;
+
+        /**
+         * Constructs an AStarHeuristic for a given graph and destination point.
+         * @param _graph the graph
+         * @param _violation_function a function that returns the violation distance for a
+         * given point
+         * @param _dest the destination point
+         */
+        explicit AStarHeuristic(const std::shared_ptr<AStarGridGraph> &_graph,
+                                const ViolationFunction &_violation_function,
+                                const Point &_dest);
+        edge_cost_t operator()(GridVertex gp);
+
+       private:
+        std::shared_ptr<AStarGridGraph> graph;
+        const Point dest_point;
+        const ViolationFunction violation_function;
+    };
+
+    struct FoundGoal
+    {
+    };
```

Comment:
What's this for?

Reply:
That's the exception that we throw when we find the goal node, I'll rename it to make it more clear what it is